idseq_pipeline/commands/non_host_alignment_functions.py

```diff
@@ -712,17 +812,81 @@ def run_gsnapl_chunk(part_suffix, remote_home_dir, remote_index_dir, remote_work
         deduplicate_m8(os.path.join(CHUNKS_RESULT_DIR, outfile_basename), os.path.join(CHUNKS_RESULT_DIR, dedup_outfile_basename))
         with iostream:
             execute_command("aws s3 cp --quiet %s/%s %s/" % (CHUNKS_RESULT_DIR, dedup_outfile_basename, SAMPLE_S3_OUTPUT_CHUNKS_PATH))
+
+    # Deduplicate multihit m8 by using taxonomy info
+    multihit_summary_file = CHUNKS_RESULT_DIR + "/summary-multihit-" + outfile_basename
+    dedup_multihit_local_outfile = CHUNKS_RESULT_DIR + "/dedup-multihit-" + outfile_basename
+    call_hits_m8(multihit_local_outfile, dedup_multihit_local_outfile, multihit_summary_file)
+    with iostream:
+        execute_command("aws s3 cp --quiet %s %s/" % (dedup_multihit_local_outfile, SAMPLE_S3_OUTPUT_CHUNKS_PATH))
+        execute_command("aws s3 cp --quiet %s %s/" % (multihit_summary_file, SAMPLE_S3_OUTPUT_CHUNKS_PATH))
+
     with iostream:
         execute_command("sed -i '$ {/^$/d;}' %s" % os.path.join(CHUNKS_RESULT_DIR, dedup_outfile_basename)) # remove blank line from end of file
     write_to_log("finished alignment for chunk %s" % chunk_id)
     return os.path.join(CHUNKS_RESULT_DIR, dedup_outfile_basename)
 
 
+def call_hits_m8(input_m8, output_m8, output_summary):
+    # Get lineage info
+    lineage_path = fetch_reference(LINEAGE_SHELF)
+    lineage_map = shelve.open(lineage_path)
+    accession2taxid_path = fetch_reference(ACCESSION2TAXID)
+    accession2taxid_dict = shelve.open(accession2taxid_path)
+    # Helper functions
+    def add_taxid_hits(accession, hits):
+        taxid = accession2taxid_dict.get(accession.split(".")[0], "NA")
+        species_taxid, genus_taxid, family_taxid = lineage_map.get(taxid, ("-100", "-200", "-300"))
+        hits["species"] += [species_taxid]
+        hits["genus"] += [genus_taxid]
+        hits["family"] += [family_taxid]
+        return hits
+    def call_hit_level(read_id, hits):
+        for i, level in enumerate(["species", "genus", "family"]):
+            taxids = hits[level]
+            taxids = [t for t in taxids if int(t) >= 0] # do not consider unclassified hits. TO DO: decide if that's an improvement
+            n = len(set(taxids)) # number of distinct taxids at this level
+            if n == 1:
+                return i+1, taxids[0]
+        return -1, -1
+    # Deduplicate m8 and summarize hits
+    read_ids = execute_command_with_output("grep -v '^#' %s | cut -f1" % input_m8).split("\n")
+    read_ids = filter(None, read_ids)
+    read_ids = set(read_ids)
+    sorted_input_m8 = input_m8 + "-sorted"
+    execute_command("sort -k1 %s > %s" % (input_m8, sorted_input_m8))
+    read_to_hit_level = {}
+    read_to_taxid = {}
+    outf = open(output_m8, "wb")
+    outf_sum = open(output_summary, "wb")
+    for read_id in read_ids:
+        # TO DO: remove inefficiency of calling grep on the entire file for each read_id.
```

Comment:
Yeah this is likely worth addressing, but first see how long it takes for biggish files.